rpcs3/util/atomic.cpp

```diff
@@ -27,6 +31,113 @@ static inline bool ptr_cmp(const void* data, std::size_t size, u64 old_value)
 	return false;
 }
 
+// Fallback implementation
+namespace
+{
+	struct waiter
+	{
+		std::condition_variable cond;
+		void* const tls_ptr;
+
+		explicit waiter(void* tls_ptr)
+			: tls_ptr(tls_ptr)
+		{
+		}
+	};
+
+	struct waiter_map
+	{
+		std::mutex mutex;
+		std::multimap<const void*, waiter> list;
+	};
+
+	// Thread's unique node to insert without allocation
+	thread_local std::multimap<const void*, waiter>::node_type s_tls_waiter = []()
+	{
+		// Initialize node from a dummy container (there is no separate node constructor)
+		std::multimap<const void*, waiter> dummy;
+		return dummy.extract(dummy.emplace(nullptr, &s_tls_waiter));
+	}();
+
+	waiter_map& get_fallback_map(const void* ptr)
+	{
+		static waiter_map s_waiter_maps[4096];
+
+		return s_waiter_maps[std::hash<const void*>()(ptr) % std::size(s_waiter_maps)];
+	}
+
+	void fallback_wait(const void* data, std::size_t size, u64 old_value)
+	{
+		auto& wmap = get_fallback_map(data);
+
+		// Update node key
+		s_tls_waiter.key() = data;
+
+		if (std::unique_lock lock(wmap.mutex); ptr_cmp(data, size, old_value))
+		{
+			// Add node to the waiter list
+			std::condition_variable& cond = wmap.list.insert(std::move(s_tls_waiter))->second.cond;
+
+			// Wait until the node is returned to its TLS location
+			while (!s_tls_waiter)
+			{
+				cond.wait(lock);
+			}
+		}
+	}
+
+	void fallback_notify(waiter_map& wmap, std::multimap<const void*, waiter>::iterator found)
+	{
+		// Return notified node to its TLS location
+		const auto ptls = static_cast<std::multimap<const void*, waiter>::node_type*>(found->second.tls_ptr);
+		*ptls = wmap.list.extract(found);
+		ptls->mapped().cond.notify_one();
+	}
+
+	void fallback_notify_one(const void* data)
+	{
+		auto& wmap = get_fallback_map(data);
+
+		std::lock_guard lock(wmap.mutex);
```

Comment:
Is `shared_mutex` (essentially a read/write lock) available on this platform? It'd allow concurrent access on `wmap.list` for reading.

Reply:
It may be available on all C++17 compatible platforms, but it's not usable with std::condition_variable, and the map is modified upon notification.

Reply:
`std::condition_variable_any` may meet the requirements for that.

Reply:
Oh, you can see that its implementation typically has a pair of std::mutex and std::condition_variable (which are only optimized to be interoperable).